### ISSUE TITLE
Being able to compute signatures using signatory.Path for a subset of the batch dimension.

### ISSUE THREAD

**Issue by jambo6** [patrick-kidger/signatory#28]
Hello,

I'd like to be able to compute signatures through the path object but only for a subset of my batch dimension at a time. At the moment, Path seems to only allow signatures to be returned on the full data before I can subset with the batch dimension. See the example below
```
path_cls = signatory.Path(data)

# Current way
signatures = path_cls.signature(i, j)
batch_signatures = signatures[batch_idxs]

# Ideal
signatures = path_cls.signature([batch_idxs, [i, j]])
```
Would it be possible to implement a behaviour like the bottom line?

Thanks!

**Comment by patrick-kidger** [comment]
Something like the bottom line can definitely be done, for some contiguous slice of batch indices.

Assuming this is for data-loading style behaviour, can you give some short example code of how you'd use your proposed 'Ideal' with a `torch.utils.data.DataLoader`, or how you're currently using the 'Current way'?

As it stands `DataLoader` tends to handle batching for you and it seems like this would involve making `Path` a replacement for `DataLoader`, which I'm not sure I'm fond of. (Making `Path` on a par with `torch.utils.data.Dataset` is probably a more elegant fit?)

**Comment by jambo6** [comment]
My use case is as follows:

I make a PyTorch dataset that consists of a `signatory.Path` class and som labels. I have a custom sampler that iterates through the batch dimension and each time selects a random partition of the dataset length. Currently, I index my dataset like
```
dataset[batch_idxs, partition] = signatory.Path(*partiton)[batch_idxs]     # This really is a stacking of all partitions 
```
Obviously it would be more memory efficient for me to be able to filter the `batch_idxs` first.

In terms of a dataset, at the core of this, I have a dataset class that I index `dataset[batch_idxs, intervals]` as it is desirable for me to extract the signature over arbitrary intervals for subsets of the batch. Again, the implementation of this for now is `signatory.Path(*interval)[batch_idxs]`.

I hope this usage makes sense!

**Comment by patrick-kidger** [comment]
So I don't understand what you mean when you write `signatory.Path(*interval)[batch_idxs]`, as that's not syntax that makes sense for `Path`.

Regardless, I think what I'm seeing here is the need for efficiently slicing and shuffling `Path`, and at that point you can wrap it up into a `Dataset` however you most prefer.

Checking the docs for `DataLoader`, I note that it provides a way of allowing you to do batching yourself, so we're not leaking an abstraction by allowing `Path` to provide batches, so my previous concern about being a bit like a `DataLoader` is probably moot.

Current plan is to implement three extra methods for `Path`: `shuffle`, `shuffle_` and `__getitem__`. The first two are out- and in-place shuffles along the batch dimension of `Path`, and can be done between epochs. The last one allows for slicing `Path` along the batch dimension so as to produce a batch for training with.

Sound good?

**Comment by patrick-kidger** [comment]
Resolved by #29 , which adds the above functionality.
